### PR TITLE
Convert CHLO TopK To LinalgExt TopK

### PR DESCRIPTION
**Pull request by KoolJBlack**
Maps a CHLO TopK To LinalgExt TopK. Additional steps during the conversion: 

- `chlo::topk` always computes `k` along the final dimension. `k` is set to the rank for `linalg_ext::topk`
- an input indices tensor is to accompany the input values is generated using a `linalg::generic` loop
- Output tensors for values and indices are initialized and filled with default values
- The `linalg_ext::topk` region is defined for a max k comparison

### CODE REVIEW

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:576]: Do we have to initialize the indices tensor? I think initializing output values is enough.

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:603]: We usually use a guard before setting insertion point. When the guard get destroyed, it'll get back to the original insertion point.

```
{
  OpBuilder::InsertionGuard guard(rewriter);
  rewriter.setInsertionPointToStart(block);
   ...
}
```

File: llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp
Comment (hanhanW) [inline llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp:1263]: `linalg_ext.yeild`?

**Review by hanhanW — CHANGES_REQUESTED** [review]
Good to see failures in integration tests. The failure means that we pass some integration tests now! Please remove `# XFAIL: *` from these two files:

https://github.com/google/iree/blob/main/integrations/tensorflow/test/iree_tf_tests/math/llvmaot__top_k.run
https://github.com/google/iree/blob/main/integrations/tensorflow/test/iree_tf_tests/math/vulkan__top_k.run

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:562]: Input indices I thought were the indices that are already set. These needn't be initialized. If these are in-out values they need to be passed to the `out` operands. Maybe I missed this on linalg ext op definition....

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:509]: This is just doing a cast? Can we fold this into the `top_k` op directly?

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:565]: Sorry for digging this up again, but can you remind me why we need a separate `indices` input tensor. That seems like information already carried by the op and does not need to be explicitly instantiated. That increases the memory foot print by a factor of 2. I know thats how the `linalg_ext.topk` is defined, but I am kind of revisiting that cause I didnt connect it till I see the lowering here.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:508]: Prefer using "assignment-like syntax" or "traditional constructor syntax (with parentheses"

```suggestion
    SmallVector<AffineMap> indexingMaps = {mapIdentity};
    SmallVector<StringRef> iterators(inputRank, "parallel");
```

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:527]: Prefer using readable methods from adaptor.

E.g., `adaptor.values().getType().cast...`, `adaptor.indices()...`.

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:534]: Looks like we don't have to re-compute `dynSizes`? They should be identical to me.

If one of batch dims is static, they should be propagated to inputs/outputs. The k dimension should always be static. In this context, we don't have to re-compute `dynSizes`.

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:502]: Should it be `indicesElementType`? They might not match.

How about asserting if `indicesElementType` is not `index` type, and use `indicesElementType` for all related tensors?

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:547]: In LLVM style, we usually declare a variable in the condition if we're going to use it. E.g.,

```cpp
if (auto intType  = valueElementType.dyn_cast<IntegerType>()) {
  ...
  intType.getWidth()
  ...
}
```

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:559]: `.result()`

**Review by hanhanW — CHANGES_REQUESTED** [review]

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp
Comment (KoolJBlack) [reply compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:565]: To clarify the arguments here: the tensor `linalg_ext::topk` needs 2 inputs and 2 outputs (though technically, they're all provided as inputs making 4 total): 
- input values and input indices
- output values and output indices

The existing `chlo::topk` op only has only a single relevant argument, the input values. See: https://github.com/tensorflow/mlir-hlo/blob/master/include/mlir-hlo/Dialect/mhlo/IR/chlo_ops.td#L768

So, for my version, I need to create the other 3 tensors and provide them as input:
- input indices- are initialized and filled using the linalg generic and index operations
- output values and indices are initialized after and filled with neg-inf and zero respectively

To answer @MaheshRavishankar  specifically, the reason I keep a separate indices tensor is to support splitting topk. I need the original indices so that I can preserve the order when merging results. I'm aware this increases memory usage. Its possible we may be able to relax this requirement for the transformer models we're working on, but at the moment we don't have concrete info that its possible. 

Does this answer your questions or am I missing something?

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp
Comment (KoolJBlack) [reply compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:562]: Dont' think I fully follow the comment? Also https://github.com/google/iree/pull/9162#discussion_r879791272 might have more context? 

The input indices are initialized with index values. Thats separate from the output tensors which both need to be provided as inputs with initialized values. Are you saying the input values tensor which is set to 0 doesn't need the `linalg.fill` operation?

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:565]: Even for splitting top-K you dont need to create a separate indices tensor. I assume here you mean splitting the reduction dimension. This comes down to tiling. At every iteration (tiled or untiled) you know the current index value. 

For example [here](https://github.com/google/iree/blob/e32cc76952d37a14f73c8a7da889edf47fcc2fce/llvm-external-projects/iree-dialects/test/Dialect/iree_linalg_ext/tiling.mlir#L83), %ARG4 is the start position of the index value for that tile. You get the index for an element by %ARG4 + index within the tile. There shouldnt be a need to compute this.

Comment (KoolJBlack) [reply compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:565]: Yes, in the tiling specific case, the loop bounds provide the index for accessing the original value. 

I haven't worked out the full process for split reduction, but the initial assumption is that the individual parallel reductions and the merge step would be separate operations and wouldn't necessarily share loop bounds or the input tensor. Once these values have been shuffled around by the parallel top k, I don't think there is any way for me to derive the original indices of the values when I merge even if I'm reusing the same input tensors. Perhaps I'm missing something?

Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:565]: This is where seeing the split of reductions as tiling helps. When you are tiling the reductions, at every step you know what your position in the untiled iteration space is by construction. So again you wouldnt need a separate operand for that.

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp
Comment (KoolJBlack) [reply compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:509]: Not following the comment, can you elaborate? Are you referring to the cast in the generic body?

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp
Comment (KoolJBlack) [reply compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:527]: These methods don't exist on the `TopKOpAdaptor`:

```
class TopKOpAdaptor {
public:
  TopKOpAdaptor(::mlir::ValueRange values, ::mlir::DictionaryAttr attrs = nullptr, ::mlir::RegionRange regions = {});

  TopKOpAdaptor(TopKOp op);

  ::mlir::ValueRange getOperands();
  std::pair<unsigned, unsigned> getODSOperandIndexAndLength(unsigned index);
  ::mlir::ValueRange getODSOperands(unsigned index);
  ::mlir::Value operand();
  ::mlir::DictionaryAttr getAttributes();
  ::mlir::IntegerAttr kAttr();
  uint64_t k();
  ::mlir::LogicalResult verify(::mlir::Location loc);
private:
  ::mlir::ValueRange odsOperands;
  ::mlir::DictionaryAttr odsAttrs;
  ::mlir::RegionRange odsRegions;
  ::llvm::Optional<::mlir::OperationName> odsOpName;
};
```

Comment (hanhanW) [reply compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp]: ah, it's reasonable because adaptor is more for accessing operands. I think these methods exist on the `TopKOp`. We're able to use `op.values().getType().cast...` etc.

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:561]: Sounds reasonable to me. The indices can be inferred in the tiling cases as well. We'll get the offsets/ivs when implementing tiling method.

Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:561]: Let me expand on this a bit with just a reduction on memrefs

```
linalg.generic {
    indexing_maps = [affine_map<(d0) -> (d0)>, affine_map<(d0) -> ()>],
    iterator_types = ["reduction"]}
    ins(%source : memref<?xf32>) outs(%zero : memref<f32>) { ... }
```

When you tile this you get

```
scf.for %iv0 = %c0 to %ub step %ts {
  %1 = subview %source[%iv0][%ts][1]
   %2 = linalg.generic ... ins(%1)
}
```

Lets lower that to loops you get

```
scf.for %iv0 = %c0 to %ub step %ts {
  %1 = subview %source[%0][%ts][1]
   scf.for %iv1 = %c0 to %ts step %c1 {
       load %1[%iv1]
   }
}
```

Within the innermost loop `%iv0 + %iv1` gives you the position in the original iteration space. This value for all iterations is what is being stored in `input_indices` and is unnecessary. You have the value always, irrespective of whether you distribute the reduction dimension or not....

Comment (ThomasRaoux) [reply compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:561]: The topk op works a bit different than a normal reduction since it needs to return the original indices of the top elements. This information would be lost when doing a tree implementation of topk. 
It is true that the index could be skipped for the first level of topk and to be efficient we should probably support it in which case we would either need to two ops or have the indices input be optional and if it is not there just relay on the loop induction variable.
We do need the existing version but if we were to also support a case where the induction variable is used that's what we would use for lowering from chlo. So it might be worth having a quick chat about it. My gut feeling is that in term of performance this is most likely going to be needed, in which case we may want to go back to the design of topk to support this.

Comment (ThomasRaoux) [reply compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:561]: That being said, I don't think it is a reason to block this PR.

Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:561]: > The topk op works a bit different than a normal reduction since it needs to return the original indices of the top elements. This information would be lost when doing a tree implementation of topk.

No it wouldnt be.

> It is true that the index could be skipped for the first level of topk and to be efficient we should probably support it in which case we would either need to two ops or have the indices input be optional and if it is not there just relay on the loop induction variable. We do need the existing version but if we were to also support a case where the induction variable is used that's what we would use for lowering from chlo.

Having two ops is definitely worth it. It just adds complexity. We definitely want to have as small as an op surface as possible. Given I am still not convinced that `input_indices` is needed, I dont see a reason for that op.

>  So it might be worth having a quick chat about it. My gut feeling is that in term of performance this is most likely going to be needed, in which case we may want to go back to the design of topk to support this.

I'd rather not have such development loops. I agree it was my mistake to not catch this on the op definition of `linalg_ext.topk`. I should have raised it at that time. Still I dont think its too late backtrack. For now the change is just a few fixes to op definition and lowering to loops/tiling (all should be minimal.

> That being said, I don't think it is a reason to block this PR.

Its a judgement call. I should have raised it in an earlier PR, but doesnt mean that we continue along this path because it was missed on initial review.

Comment (KoolJBlack) [reply compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:561]: @ThomasRaoux @MaheshRavishankar and @KoolJBlack discussed this further offline. The decision has been to make the `%input_indices` variable of TopK optional. I've created an issue for this task: https://github.com/google/iree/issues/9204

With this in mind, I think we can resolve this conversation for now. Since its not specifically blocking here,  I propose we commit this PR so that we'll have an e2e implementation of Topk. I will followup with a PR modifying the entire e2e implementation with the additional optional use case.

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:486]: Why `RankedTensorType` here vs. `ShapedType` everywhere below.?

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:561]: My (strong) preference here is that this PR be held off until the changes to the semantics of `linalg_ext.topk` are made and all the transformations with the changes are verified. This PR would need to change to not generate the `linalg.generic` that is just materializing indices. Given that this PR is essentially generating two ops, and one of them is unnecessary, it seems better to land this after all changes to the `linalg_ext.topk`.
Since Kojo mentioned that it will be easier for him to make progress on fixing things up after this PR lands, I am not going to block it, but landing this seems premature to me.

Comment (hanhanW) [reply compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:561]: We're exploring how it works at this stage, so making it optional sounds good to me. We might get some perf win if the shape is static. Because we don't have to compute them during execution. 

It's always easier to me when things not getting complicated. After thinking a while, it's fine to land the PR firstly because I assume that we only need small changes when adding optional support. E.g., adding an if-else statement to decide whether we need indices operand. I'll review it later today and let Kojo make the call.

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:495]: Move it after `indicesElementType` declaration, so we can use `indicesElementType` variable here.

Also, lets raise some failure messages and move the comment to there.

```
return rewriter.notifyMatchFailure(op, "...");
```

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:516]: `mapIdentity` is only used once. Maybe combine them into one statement or push_back `rewriter.getMultiDimIdentityMap(inputRank)` directly ?

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:558]: I still think that we don't have to initialize output indices...

I think the assumption is that input values are all non-negative. In this context, initializing output values is enough. Because top_k compares the values firstly. The indices are considered only when values are equal.

Without the assumption, this is still wrong. Say that if all the inputs are negative. We'll return K `(neg ,0)` pairs?

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:567]: style nit

```suggestion
    SmallVector<Location> locations(2, loc);
```

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp
Comment (KoolJBlack) [reply compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:561]: I've gone ahead and sent out https://github.com/google/iree/pull/9228. I think this change is fairly straightforward, so if you want to review that first, I can commit that then this.

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp
Comment (KoolJBlack) [reply compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:558]: to clarify, we're not making any assumptions about the input. Values can be negative or positive. The choice of `negInf` is to ensure any value higher than the lowest possible value is chosen, and if all the values happen to also be `negInf` then they'll fall to the second case where the "first" neg infinity is chosen. 

Actually, now that you point this out, I realize that the `zeroTensor` assigned to the indices is actually incorrect; it should be `maxInf` instead so that if all the values are negative infinity, the first occurrences are chosen. I've updated the code to reflect this. 

Ultimately, as stated above, I believe both these values do need to be initialized. Also as Mahesh stated, I do believe if the values aren't initialized, they are essentially undefined, which will lean do undefined behavior with these comparisons.

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp
Comment (KoolJBlack) [reply compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:539]: https://github.com/google/iree/pull/9228 has merged and this change has been updated to on use a single input.

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:528]: Should be consistent. If you really want to use `i32` type, we can declare a `i32Type` variable and check if `indicesElementType == i32Type` in the beginning.

```suggestion
    Attribute posInfAttr = rewriter.getIntegerAttr(
        indicesElementType, APInt::getSignedMaxValue(32));
```

**Review by hanhanW — APPROVED** [review]
LGTM

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp]: Although I'm not really convinced that the initialization is needed, I think it shouldn't be a blocker for landing the PR.

The topk operation returns the indices. Since the ivs and k are known, we should be able to check what values to use. It can be checked by comparing ivs and k to me. Maybe the current implementation puts some limits on it.

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:515]: Comment seems to be off from implementation below.

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:529]: If I am reading this correctly (and following Hanhans comments), we should not need to initialize the indices tensor, unless this is some runtime error mode that is being signalled. This would allow the `fill` op for the output indices to be dropped.

**Review by MaheshRavishankar — APPROVED** [review]
Left a couple of minor comments on this PR. I think its fine to land this. Have been thinking about the disconnect here with the input indices being made optional (something about that didnt sit well with me). I think I have a solution that is more uniform. Ill file that as a separate issue.

**Pull request merged**
